### ISSUE TITLE
Don't assume functions starting with _ are ES5 constructors

### ISSUE THREAD

**Issue by kevinoid** [eslint/eslint#6085]
**What version of ESLint are you using?**

2.3.0 or later (any version including 13c7c19f).

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```yml
rules:
  consistent-return: 2
```

**What did you do? Please include the actual source code causing the issue.**

```js
function _doSomething(condition) {
    if (condition) {
        return true;
    }
}
```

**What did you expect to happen?**

```
  1:10  error  Expected to return a value at the end of this function  consistent-return
```

**What actually happened? Please include the actual, raw output from ESLint.**

No output is produced.

This is an unintended result of my change in #5398 which excludes functions which pass `isES5Constructor`.  The logic in `isES5Constructor` checks that [the first character of the function name is invariant under `.toLocaleUpperCase`](https://github.com/eslint/eslint/blob/v2.9.0/lib/ast-utils.js#L60), which includes both upper-case and caseless characters (including punctuation).  This is especially unfortunate since `_` is commonly used to indicate semi-private methods.

I suggest excluding `_` (and maybe `$`) from `isES5Constructor`.  In the cases where a constructor begins with these characters, the user can add inline config to disable `consistent-return` and `no-invalid-this` as needed.  If that sounds agreeable, I'll work up a PR.

Thoughts?

CC: @ilyavolodin, @mysticatea, @nzakas since you guys helped with #5398.

**Comment by nzakas** [comment]
That definitely sounds like a bug, I just wonder if the better fix is upstream to the `esutil` package rather than fixing in ESLint.

**Comment by kevinoid** [comment]
That's a good thought.  It seems like most of the current functions in `esutils` are objective while `isES5Constructor` is a bit more subjective and convention-based.  If that's ok, it sounds like a good place to fix it.

**Comment by nzakas** [comment]
Yeah, I'd say start there and see where you get. We can leave this open to track it on our side until we know what's going to happen.

**Comment by nzakas** [comment]
Sorry, I misunderstood the code. `isES5Constructor` doesn't come from another package, it comes from us, so we should just update it.

**Comment by kevinoid** [comment]
No worries.  That sounds good.  I'll send a PR shortly.

**Comment by mysticatea** [comment]
Wow, good catch!
Thank you!
